src/native/containers/dn-array.h

```diff
@@ -0,0 +1,74 @@
+
+#ifndef __DN_ARRAY_H__
+#define __DN_ARRAY_H__
+
+#include "dn-utils.h"
+
+typedef struct _dn_array_t dn_array_t;
+struct _dn_array_t {
+	char *data;
+	int32_t len;
+};
+
+dn_array_t *
+dn_array_new (
+	bool zero_terminated,
+	bool clear_,
+	uint32_t element_size);
+
+dn_array_t *
+dn_array_sized_new (
+	bool zero_terminated,
+	bool clear_,
+	uint32_t element_size,
+	uint32_t reserved_size);
+
+char *
+dn_array_free (
+	dn_array_t *array,
+	bool free_segment);
+
+void
+dn_array_free_segment (char *segment);
+
+dn_array_t *
+dn_array_append_vals (
+	dn_array_t *array,
+	const void *data,
+	uint32_t len);
+
+dn_array_t *
+dn_array_insert_vals (
+	dn_array_t *array,
+	uint32_t index_,
+	const void *data,
+	uint32_t len);
+
+dn_array_t *
+dn_array_remove_index (
+	dn_array_t *array,
+	uint32_t index_);
+
+dn_array_t *
+dn_array_remove_index_fast (
+	dn_array_t *array,
+	uint32_t index_);
```

Comment:
Accepting an unsigned index but having a signed length just makes for compiler warnings and casting nits. I think we should decide on what semantics we want here and be consistent about our length and index types.

Reply:
Jupp, that is an inhrited problem with the glib containers used in Mono, but if we decide to break glib API combability the API will amke sure index and length will be matching unsigned types.